rocr: Add support for Mipmapped Array

ApurvMishra-amd opened the pull request:
## Motivation

Add support for Mipmapped Array in ROCm Runtime

## Technical Details

This PR adds support for Mipmapped Array functionality to the ROCm Runtime, enabling GPU applications to work with multi-level texture mipmaps. The implementation introduces new public APIs for creating, querying, and managing mipmapped arrays across different GPU architectures.

## Test Plan

Tests being added in 'rocrtst' test suite for validation.

ApurvMishra-amd committed:
rocr: Add support for Mipmapped Array

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager.cpp:
```diff
@@ -118,6 +118,56 @@ void Sampler::Destroy(const Sampler* sampler) {
+
+  // Find a GPU memory pool for image data
+  hsa_status_t status = hsa_amd_agent_iterate_memory_pools(
+                        agent, [](hsa_amd_memory_pool_t p, void* data) {
```
Not sure why this is different from Image* Image::Create() which doesn't call hsa_amd_agent_iterate_memory_pools.

shwetagkhatri replied:
Thanks for pointing this out @TomSang. We should be allocating Image data from the kernarg_pool itself. i will update this part of the code.

ShwetakAMD committed:
rocr: Adding Mipmap Support: Images and Mipmap API unification, and fixed code review comments

Introduced unified V2 APIs for simplified image and mipmap management and addressed
critical memory safety issues
Mipmap V2 APIs:
- Added hsa_ext_image_descriptor_v2_t with mipmap_levels field
- Implemented hsa_ext_image_create_v2(), hsa_ext_image_destroy_v2(), and
  hsa_ext_image_data_get_info_v2() for unified image/mipmap management
- V2 APIs route internally based on mipmap_levels: >1 uses mipmap path, else standard path
- Zero breaking changes: V1 APIs remain unchanged, V2 APIs are additive

ShwetakAMD committed:
rocr: Navi Mipmap support: Images and Mipmap API unification, and updated GFX11/12 logic

Introduced unified V2 APIs for simplified image and mipmap management and addressed
critical memory safety issues
Mipmap V2 APIs:
- Added hsa_ext_image_descriptor_v2_t with mipmap_levels field
- Implemented hsa_ext_image_create_v2(), hsa_ext_image_destroy_v2(), and
  hsa_ext_image_data_get_info_v2() for unified image/mipmap management
- V2 APIs route internally based on mipmap_levels: >1 uses mipmap path, else standard path
- Zero breaking changes: V1 APIs remain unchanged, V2 APIs are additive

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_ai.cpp:
```diff
@@ -611,5 +614,320 @@ uint32_t ImageManagerAi::GetAddrlibSurfaceInfoAi(
+    sq_img_rsrc_word2_u word2;
+    sq_img_rsrc_word3_u word3;
+    sq_img_rsrc_word4_u word4;
+    sq_img_rsrc_word5_u word5;
+    sq_img_rsrc_word6_u word6;
+    sq_img_rsrc_word7_u word7;
+
+    ADDR2_COMPUTE_SURFACE_INFO_OUTPUT out = {0};
+    uint32_t swizzleMode = GetAddrlibSurfaceInfoAi(
+                        mipmap.component, mipmap.desc, mipmap.num_levels,
+                        mipmap.tile_mode, mipmap.row_pitch, mipmap.slice_pitch, out);
+    if (swizzleMode == (uint32_t)(-1)) {
+      return HSA_STATUS_ERROR;
+    }
+    mipmap.addr_output.addr2 = out;
```
Looks like out.pMipInfo is not allocated . Will it be used ?

shwetagkhatri replied:
Heap allocation with proper cleanup added.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx11.cpp:
```diff
@@ -641,7 +641,6 @@ uint32_t ImageManagerGfx11::GetAddrlibSurfaceInfoNv(
   in.width = width;
   in.height = height;
   in.numSlices = num_slice;
```
Do we need in.numMipLevels = num_mipmap_levels;?

shwetagkhatri replied:
Yes.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx11.cpp:
```diff
@@ -810,5 +809,325 @@ hsa_status_t ImageManagerGfx11::FillImage(const Image& image, const void* patter
+    SQ_IMG_RSRC_WORD5 word6;
+    SQ_IMG_RSRC_WORD5 word7;
+
+    ADDR2_COMPUTE_SURFACE_INFO_OUTPUT out = {0};
```
Looks like out.pMipInfo is not allocated . Will it be used ?

shwetagkhatri replied:
Ywes. Allocated and fixed this issue.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx12.cpp:
```diff
@@ -216,12 +216,19 @@ ImageManagerGfx12::~ImageManagerGfx12() {}
+  mip_info_storage.resize(num_mipmap_levels);
 
+  // Assign pointer to allocated memory
+  out.pMipInfo = mip_info_storage.empty() ? nullptr : mip_info_storage.data();
```
See Copilot's comment

shwetagkhatri replied:
Added proper heap allocation with proper cleanup.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx12.cpp:
```diff
@@ -664,54 +669,52 @@ uint32_t ImageManagerGfx12::GetAddrlibSurfaceInfoNv(
   in.width = width;
   in.height = height;
   in.numSlices = num_slice;
-  in.pitchInElement = image_data_row_pitch / image_prop.element_size;
```
Do we still need  in.pitchInElement = image_data_row_pitch / image_prop.element_size;?

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx12.cpp:
```diff
@@ -793,6 +796,9 @@ uint32_t ImageManagerGfx12::GetAddrlibSurfaceInfoNv(
         ADDR3_COMPUTE_SURFACE_INFO_OUTPUT localOut = {0};
+        std::vector<ADDR3_MIP_INFO> mip_info_storage;
+        mip_info_storage.resize(num_mipmap_levels);
+        localOut.pMipInfo = mip_info_storage.empty() ? nullptr : mip_info_storage.data();
```
See concern of Coplilot

shwetagkhatri replied:
Fixed.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx12.cpp:
```diff
@@ -908,5 +914,349 @@ hsa_status_t ImageManagerGfx12::FillImage(const Image& image, const void* patter
+    mip_info_storage.resize(mipmap.num_levels);
+
+    // Assign pointer to allocated memory
+    out.pMipInfo = mip_info_storage.empty() ? nullptr : mip_info_storage.data();
```
Same concern

shwetagkhatri replied:
Did proper heap allocation with proper cleanup.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_nv.cpp:
```diff
@@ -804,5 +802,336 @@ hsa_status_t ImageManagerNv::FillImage(const Image& image, const void* pattern,
+    SQ_IMG_RSRC_WORD2 word2;
+    SQ_IMG_RSRC_WORD3 word3;
+    SQ_IMG_RSRC_WORD4 word4;
+    SQ_IMG_RSRC_WORD5 word5;
+    SQ_IMG_RSRC_WORD5 word6;
+    SQ_IMG_RSRC_WORD5 word7;
+
+    ADDR2_COMPUTE_SURFACE_INFO_OUTPUT out = {0};
+    uint32_t swizzleMode = GetAddrlibSurfaceInfoNv(
+                        mipmap.component, mipmap.desc, mipmap.num_levels,
+                        mipmap.tile_mode, mipmap.row_pitch, mipmap.slice_pitch, out);
+    if (swizzleMode == (uint32_t)(-1)) {
+      return HSA_STATUS_ERROR;
+    }
+    mipmap.addr_output.addr2 = out;
```
See Copilot's comment

shwetagkhatri replied:
Done.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_runtime.cpp:
```diff
@@ -574,6 +668,142 @@ hsa_status_t ImageRuntime::DestroySamplerHandle(
+  uint32_t major_ver = MajorVerFromDevID(chip_id);
+  if (major_ver < 9) {
+    debug_print("ERROR: Mip level views not supported on GFX%u hardware\n", major_ver);
+    return HSA_STATUS_ERROR_INVALID_ARGUMENT;
```
Same concern

shwetagkhatri replied:
Return value HSA_STATUS_ERROR_INVALID_ARGUMENT is correct.

JakeMerdichAMD commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_runtime.cpp:
```diff
@@ -574,6 +668,164 @@ hsa_status_t ImageRuntime::DestroySamplerHandle(
+    mipmap_array->tile_mode = Image::TileMode::LINEAR;
+  } else {
+    Image::TileMode tileMode =
+    (profile == HSA_PROFILE_BASE && mipmap_descriptor.geometry != HSA_EXT_IMAGE_GEOMETRY_1DB)
```
What's the reasoning behind the HSA_PROFILE_BASE restriction? Wouldn't this force linear for all images on linux? That usually comes with a performance penalty.

JakeMerdichAMD commented on projects/rocr-runtime/runtime/hsa-runtime/image/resource.h:
```diff
@@ -202,6 +203,68 @@ typedef struct Sampler {
+  } else if (addr_output.addr3.pMipInfo) {
+    delete[] addr_output.addr3.pMipInfo;
+    addr_output.addr3.pMipInfo = nullptr;
+  }
```
I'm not sure the if check here does what you want. If addr3 is the active member of the union, wouldn't the addr2.pMipInfo check test random garbage in the addr3 struct for nullness?

TomSang replied:
It should check which addr lib version is used

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_nv.cpp:
```diff
@@ -635,9 +635,7 @@ uint32_t ImageManagerNv::GetAddrlibSurfaceInfoNv(
-  // Custom Pitch is supported in gfx1030 and beyond
-  if (minor_ver >= 3)
-    in.pitchInElement = image_data_row_pitch / image_prop.element_size;
+
```
missing 
  in.numMipLevels = num_mipmap_levels;

JakeMerdichAMD commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_gfx12.cpp:
```diff
@@ -793,13 +804,19 @@ uint32_t ImageManagerGfx12::GetAddrlibSurfaceInfoNv(
+
+        // Allocate temporary memory for mip info (freed after use)
+        localOut.pMipInfo = new ADDR3_MIP_INFO[num_mipmap_levels];
+        memset(localOut.pMipInfo, 0, sizeof(ADDR3_MIP_INFO) * num_mipmap_levels);
```
If you're not going to use any of the per-mip image info this can be set to null and it will be ignored. Ditto on all chip versions.

TomSang replied:
We need 
hsa_status_t HSA_API hsa_amd_mipmap_array_get_level(hsa_agent_t agent,   const hsa_ext_image_t* mipmapped_array,
uint32_t mip_level,    hsa_ext_image_t* level_image_out) 

Looks like  pMipInfo can  be used for it,  but it isn't referenced really.    
Not sure whether current implementation of   hsa_amd_mipmap_array_get_level() can work well.

ApurvMishra-amd replied:
Initially, I had implemented the _hsa_ext_image_mipmap_array_get_level_ API using **pMipInfo** but that implementation seemed elaborate. Then, as per Jake's suggestion, I moved to using the BASE_LEVEL and LAST_LEVEL fields of SRD to extract the single MIP level info. Does that seem correct?
@JakeMerdichAMD @TomSang

TomSang replied:
Looks like  hsa_ext_image_mipmap_array_get_level API is implemented through pMipInfo is the right path. If using BASE_LEVEL and LAST_LEVEL fields of SRD to extract the single MIP level info,  we cannot create another view on top of the leveled image.   But that is required when we create a texture object.  See https://github.com/ROCm/rocm-systems/blob/develop/projects/clr/hipamd/src/hip_texture.cpp#L253  that creates another view on top of the leveled image.   So we need using pMipInfo to implement hsa_ext_image_mipmap_array_get_level

JakeMerdichAMD reviewed (commented):
A few minor notes, but this generally looks good.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/image/resource.h:
```diff
@@ -202,6 +203,68 @@ typedef struct Sampler {
+    tile_mode = LINEAR;
+  }
+
+~MipmappedArray() {
```
should   ~Image() {} be virtual?
Looks like Image::Destroy(const Image* image) need that to destroy MipmappedArray

shwetagkhatri replied:
Yes! Thank you for catching this critical bug ..Other wise Image::Destroy() was only calling ~Image() even when we were passing MipmappedArray*  which is the derived class that needs to be destroyed.  I have applied this fix and am ready for my next upload.

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/hsacore.so.def:
```diff
@@ -260,8 +260,10 @@ global:
 	hsa_amd_portable_export_dmabuf_v2;
-	hsa_amd_ais_file_write;
-	hsa_amd_ais_file_read;
+	hsa_amd_mipmap_array_get_info;
```
should be 
  hsa_ext_image_mipmap_array_get_level
  hsa_ext_image_create_v2
  hsa_ext_image_data_get_info_v2
  hsa_ext_image_destroy_v2
The same in runtime/hsa-runtime/hsacore.dll.def

TomSang replied:
These haven't been added in rocr-runtime/runtime/hsa-runtime/hsacore.dll.def, 
hsa_ext_image_mipmap_array_get_level
hsa_ext_image_create_v2
hsa_ext_image_data_get_info_v2
hsa_ext_image_destroy_v2

dayatsin-amd replied:
This is removing the 2 existing APIs:
hsa_amd_ais_file_write;
hsa_amd_ais_file_read;

TomSang commented on projects/rocr-runtime/runtime/hsa-runtime/inc/hsa_ext_image.h:
```diff
@@ -864,6 +956,80 @@ hsa_status_t HSA_API hsa_ext_image_destroy(
+ * @retval ::HSA_STATUS_ERROR_INVALID_ARGUMENT @p image is invalid.
+ */
+hsa_status_t HSA_API hsa_ext_image_destroy_v2(hsa_agent_t agent, hsa_ext_image_t image);
+
```
Need add

hsa_status_t HSA_API hsa_ext_image_mipmap_array_get_level(hsa_agent_t agent,
                                                          const hsa_ext_image_t* mipmapped_array,
                                                          uint32_t mip_level,
                                                          hsa_ext_image_t* level_image_out);

ApurvMishra-amd committed:
rocr: Add pMipInfo for mip level view API

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

ApurvMishra-amd committed:
rocr: Add mip level extraction API to hsa_ext_image

Add hsa_ext_image_mipmap_array_get_level to hsa_ext_image APIs
Add new image-mipmap integrated v2 APIs to hsacore dll

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp:
```diff
@@ -0,0 +1,194 @@
+}
+
+
+// Public API implementations for mipmapped array query & creation
```
These can be inside existing hsa_ext_image.* files

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/inc/hsa_amd_mipmap_impl.h:
```diff
@@ -0,0 +1,89 @@
+
+hsa_status_t GetImageExtensionTable(hsa_ext_images_1_pfn_t* table);
+
+extern "C" {
```
These can be inside existing hsa_ext_image.* files

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp:
```diff
@@ -0,0 +1,194 @@
+    uint32_t requested_levels,
+    hsa_amd_mipmap_array_info_t* info) {
+
+    if (!desc || !info) return HSA_STATUS_ERROR_INVALID_ARGUMENT;
```
It looks like you can call GetMipmapArraySizeAndAlignment directly as GetMipmapArraySizeAndAlignment also sets size, alignment, max_levels, levels_used to 0.

i.e:

return rocr::image::ImageRuntime::instance()->GetMipmapArraySizeAndAlignment(
        agent, *desc, info->levels_used,
        HSA_EXT_IMAGE_DATA_LAYOUT_OPAQUE,  // Always tiled for mipmaps
        0, 0,                              // row_pitch and slice_pitch not used for tiled layout
        info->size, info->alignment, info->max_levels);

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp:
```diff
@@ -0,0 +1,194 @@
+    uint32_t num_mipmap_levels,
+    hsa_ext_image_t* image_out) {
+
+    TRY
```
I think you need ; after the TRY and CATCH

dayatsin-amd replied:
Still missing ;

ApurvMishra-amd replied:
This file does not exist anymore, but the same suggestion has been applied to _hsa_ext_image.cpp_ where these methods were moved.

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp:
```diff
@@ -0,0 +1,194 @@
+
+        debug_print("Asserted: num_mipmap_levels <= max_levels");
+
+        auto* rt = rocr::image::ImageRuntime::instance();
```
We should throw exception if rocr::image::ImageRuntime constructor fails. And then assume rocr::image::ImageRuntime::instance() is through the life of the process

return rocr::image::ImageRuntime::instance()->CreateMipmapArrayHandle(agent, *desc,
            image_data, access_permission, num_mipmap_levels,
            HSA_EXT_IMAGE_DATA_LAYOUT_OPAQUE, 0, 0, *image_out);

dayatsin-amd replied:
This comment was not addresssed

ApurvMishra-amd replied:
Okay, I added this to _hsa_ext_image.cpp_ where these methods were moved.

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp:
```diff
@@ -0,0 +1,194 @@
+hsa_amd_mipmap_array_destroy(const hsa_ext_image_t* image) {
+
+  TRY
+  if (!image) return HSA_STATUS_ERROR_INVALID_ARGUMENT;
```
if (!image || image->handle == 0) return HSA_STATUS_ERROR_INVALID_ARGUMENT;

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp:
```diff
@@ -0,0 +1,194 @@
+                                                          const hsa_ext_image_t* mipmapped_array,
+                                                          uint32_t mip_level,
+                                                          hsa_ext_image_t* level_image_out) {
+  TRY if (!mipmapped_array || !level_image_out) { return HSA_STATUS_ERROR_INVALID_ARGUMENT; }
```
Please move if statement on a new line

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +388,138 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+    desc_v1.height = image_descriptor->height;
+    desc_v1.depth = image_descriptor->depth;
+    desc_v1.array_size = image_descriptor->array_size;
+    desc_v1.format = image_descriptor->format;
```
Duplicate code between the if/else section. We can move this section above the "if (mipmap_levels > 1)"

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +388,138 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+    desc_v1.array_size = image_descriptor->array_size;
+    desc_v1.format = image_descriptor->format;
+
+    hsa_status_t status = ImageRuntime::instance()->GetMipmapArraySizeAndAlignment(
```
return ImageRuntime::instance()->GetMipmapArraySizeAndAlignment(
        agent, desc_v1, mipmap_levels, HSA_EXT_IMAGE_DATA_LAYOUT_OPAQUE, 0, 0, 
image_data_info->size,
image_data_info->alignment, max_levels_out);

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +388,138 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+                                                          const hsa_ext_image_t* mipmapped_array,
+                                                          uint32_t mip_level,
+                                                          hsa_ext_image_t* level_image_out) {
+  TRY if (!mipmapped_array || !level_image_out) { return HSA_STATUS_ERROR_INVALID_ARGUMENT; }
```
if statement on new line

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_nv.cpp:
```diff
@@ -804,5 +808,354 @@ hsa_status_t ImageManagerNv::FillImage(const Image& image, const void* pattern,
+
+    // Allocate persistent mip info storage for level extraction
+    ADDR2_MIP_INFO* mip_info_storage = new ADDR2_MIP_INFO[mipmap.num_levels];
+    memset(mip_info_storage, 0, sizeof(ADDR2_MIP_INFO) * mipmap.num_levels);
```
You can add {} to zero the array:

ADDR2_MIP_INFO* mip_info_storage = new ADDR2_MIP_INFO[mipmap.num_levels]**{}**;

dayatsin-amd replied:
Comment not addressed

ApurvMishra-amd replied:
The code was changed.

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +388,138 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+      (image_descriptor->mipmap_levels == 0) ? 1 : image_descriptor->mipmap_levels;
+
+  // Convert V2 descriptor to V1 for internal use
+  hsa_ext_image_descriptor_t desc_v1;
```
Initialize to 0:
hsa_ext_image_descriptor_t desc_v1 = {};

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +388,138 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+    uint32_t max_levels_out = 0;
+
+    // Convert V2 descriptor to V1 for internal use
+    hsa_ext_image_descriptor_t desc_v1;
```
Initliaze to 0:
hsa_ext_image_descriptor_t desc_v1 = {};

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/image_manager_nv.cpp:
```diff
@@ -804,5 +808,354 @@ hsa_status_t ImageManagerNv::FillImage(const Image& image, const void* pattern,
+    ADDR2_COMPUTE_SURFACE_INFO_OUTPUT out = {0};
+
+    // Allocate persistent mip info storage for level extraction
+    ADDR2_MIP_INFO* mip_info_storage = new ADDR2_MIP_INFO[mipmap.num_levels];
```
Can we use a std::shared_ptr here instead of new/delete for the array?

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/resource.h:
```diff
@@ -202,6 +207,72 @@ typedef struct Sampler {
+/// @brief Structure representing a mipmapped image array.
+typedef struct MipmappedArray : public Image {
+private:
+  MipmappedArray() {
```
Can you initialize the class elements in the "pre-constructor":
 MipmappedArray(): 
   component.handle(0), data(nullptr), size(0) etcc {

dayatsin-amd replied:
data, row_pitch and slice_pitch should be initialized in pre-constructor.

ApurvMishra-amd replied:
Added _data_, _row_pitch_ and _slice_pitch_ in Image pre-contructor as _MipmappedArray_ struct simply derives these from Image struct.

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/resource.h:
```diff
@@ -202,6 +207,72 @@ typedef struct Sampler {
+public:
+ /// @brief Create a MipmappedArray.
+ /// Only internal metadata is allocated; image data must be provided by the user.
+ static MipmappedArray* Create(hsa_agent_t agent);
```
Could this be a constructor? 

MipmappedArray(hsa_agent_t agent);

ApurvMishra-amd replied:
This is not a constructor, but an initial create function to define a mipmapped array and allocate memory.

dayatsin-amd replied:
I know it is now currently a constructor :) I am asking to make it a constructor.
That way you can have a constructor and destructor for MipmappedArray instead of Create/Destroy function calls.

ApurvMishra-amd replied:
Okay, I had just designed it as per the existing Image class structure. I am making a note of it to incorporate this in a later commit.

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/resource.h:
```diff
@@ -202,6 +207,72 @@ typedef struct Sampler {
+ static MipmappedArray* Create(hsa_agent_t agent);
+
+ /// @brief Destroy a MipmappedArray.
+ static void Destroy(const MipmappedArray* array);
```
Can this be the combined with the MipmappedArray desctructor?

ApurvMishra-amd replied:
Yes

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/inc/hsa_amd_mipmap.h:
```diff
@@ -0,0 +1,196 @@
+////////////////////////////////////////////////////////////////////////////////
```
Please move to hsa_ext_image.h

Otherwise, if you want to add new files, we got instructions from Legal to use MIT license:

************************************************************************************************************************
*
*  Copyright (C) Advanced Micro Devices, Inc.  All rights reserved.
*  SPDX-License-Identifier: MIT
*
***********************************************************************************************************************/

dayatsin-amd reviewed (changes requested):
Generally looks good. Some comments.

ApurvMishra-amd committed:
rocr: Add image-mipmap integrated APIs with given layout

Add new APIs for creation of images and getting required info
which integrates mipmap functionality as well, and is called
with a pre-determined image layout

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

ApurvMishra-amd committed:
Revert "rocr: Add pMipInfo for mip level view API"

This reverts commit b19e9be4b2eb5d9c15ad0a9315210deb01d0f9eb.

Use BASE/LAST LEVEL to get specfic MIP level info

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

ApurvMishra-amd committed:
rocr: Multiple code changes to mipmap implementation

Remove hsa_amd_mipmap code after image-mipmap integration
Add new APIs to hsacore.so.def list
Zero-initialize v2 image descriptors
Add max_mip definition to GFX versions below 12
Call GetMipmapArraySizeAndAlignment directly for info
Copy parent mipmapped array structure to level_view
Initialize MipmappedArray elements in pre-constructor

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

ApurvMishra-amd committed:
rocr: Add wrapper functions for mipmapped arrays

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

TomSang committed:
rocr: if tiling mode isn't linear, let addr lib decide the best mode.

TomSang commented:
!verify

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +430,156 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+                                                          const hsa_ext_image_t* mipmapped_array,
+                                                          uint32_t mip_level,
+                                                          hsa_ext_image_t* level_image_out) {
+  TRY
```
TRY/CATCH needs semi colon

cfreeamd replied:
@dayatsin-amd  I don't think you need a ";" after TRY and CATCH. I may be misunderstanding. This is what I see as the definition of those:

#define TRY try {
#define CATCH } catch(...) { return AMD::handleException(); }
#define CATCHRET(RETURN_TYPE) } catch(...) { return AMD::handleExceptionT<RETURN_TYPE>(); }

Am I missing something?

cfreeamd replied:
Though I do see that pattern elsewhere...

dayatsin-amd replied:
Yes, I don't think it is absolutely necessary to put the semi-colons, but I want to be consistent

ApurvMishra-amd replied:
Done!

ApurvMishra-amd committed:
rocr: Minor modifications to the mipmap code

Add semi-colon after TRY/CATCH consistently,
Throw throw exception if ImageRuntime instance fails,
Add initializations in struct defintion.

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

ApurvMishra-amd commented:
!verify

dayatsin-amd commented on projects/rocr-runtime/runtime/hsa-runtime/image/hsa_ext_image.cpp:
```diff
@@ -388,6 +430,162 @@ hsa_status_t hsa_amd_image_create(hsa_agent_t agent,
+  TRY;
+  if (!mipmapped_array || !level_image_out) { return HSA_STATUS_ERROR_INVALID_ARGUMENT; }
+
+  auto* rt = rocr::image::ImageRuntime::instance();
```
return rocr::image::ImageRuntime::instance()->GetMipmapArrayLevelHandle(agent, *mipmapped_array, mip_level, *level_image_out);

ApurvMishra-amd replied:
I changed ImageRuntime::CreateSingleton() to throw AMD::hsa_exception with descriptive error messages when initialization fails, rather than returning NULL.

ApurvMishra-amd committed:
rocr: Throw exception on initialization failure

Throw AMD::hsa_exception with descriptive error messages
when initialization fails, rather than returning NULL.

Signed-off-by: Apurv Mishra <Apurv.Mishra@amd.com>

ApurvMishra-amd commented:
!verify

ApurvMishra-amd commented:
!verify

ApurvMishra-amd committed:
Merge branch 'develop' into SWDEV-539526/ROCR-HIP-Mipmaps-Integration
